midp/localmsg.js

```diff
@@ -78,9 +87,16 @@ LocalMsgConnection.prototype.getServerMessage = function(data) {
 }
 
 LocalMsgConnection.prototype.waitServerMessage = function(data) {
+    var ctx = $.ctx;
+
     asyncImpl("I", new Promise((function(resolve, reject) {
-        this.serverWaiting.push(function() {
-            resolve(this.getServerMessage(data));
+        this.serverWaiting.push(function(successful) {
+            if (successful) {
+                resolve(this.getServerMessage(data));
+            } else {
+                ctx.setAsCurrentContext();
+                reject($.newIOException("Client disconnected"));
```

Comment:
I like the new design of sending success/failure across the wire as the first part of the protocol :)